week7/answers.sql

```diff
@@ -1,3 +1,183 @@
+JOIN items i ON eq.item_id = i.item_id;
+
+CREATE VIEW team_items AS
+SELECT tm.team_id, t.name AS team_name, i.name AS item_name, i.armor, i.damage
+FROM teams t
+JOIN team_members tm ON t.team_id = tm.team_id
+JOIN characters ch ON tm.character_id = ch.character_id
+JOIN inventory inv ON ch.character_id = inv.character_id
+JOIN items i ON inv.item_id = i.item_id
+UNION
+SELECT tm.team_id, t.name AS team_name, i.name AS item_name, i.armor, i.damage
+FROM teams t
+JOIN team_members tm ON t.team_id = tm.team_id
+JOIN characters ch ON tm.character_id = ch.character_id
+JOIN equipped eq ON ch.character_id = eq.character_id
+JOIN items i ON eq.item_id = i.item_id;
+
+DELIMITER //
+
+CREATE FUNCTION armor_total(character_id INT) RETURNS INT
+READS SQL DATA
+BEGIN
+    DECLARE total_armor INT;
+    
+    SELECT COALESCE(SUM(cs.armor), 0) + COALESCE(SUM(i.armor), 0) AS total
+    INTO total_armor
+    FROM character_stats cs
+    JOIN equipped eq ON cs.character_id = eq.character_id
+    JOIN items i ON eq.item_id = i.item_id
+    WHERE cs.character_id = character_id;
+    
+    RETURN total_armor;
+END;
+//
+
+CREATE PROCEDURE attack(IN attacked_character_id INT, IN attacking_item_id INT)
+BEGIN
+    DECLARE total_armor INT;
+    DECLARE total_damage INT;
+    
+    SET total_armor = armor_total(attacked_character_id);
+    
+    SELECT damage INTO total_damage
+    FROM items
+    WHERE item_id = attacking_item_id;
+    
+    SET total_damage = total_damage - total_armor;
+    
+    IF total_damage > 0 THEN
+        UPDATE character_stats
+        SET health = CASE
+            WHEN health - total_damage > 0 THEN health - total_damage
+            ELSE 0
+            END
+        WHERE character_id = attacked_character_id;
+        
+        DELETE FROM characters
+        WHERE character_id = attacked_character_id AND health <= 0;
+        
+        -- Assuming cascading deletes are enabled for character-related tables to delete their possessions and team membership
+    END IF;
+END;
+//
+
+CREATE PROCEDURE equip(IN inventory_item_id INT)
+BEGIN
+    DECLARE equipped_item_id INT;
+    
+    SELECT item_id INTO equipped_item_id
+    FROM inventory
+    WHERE inventory_id = inventory_item_id;
+    
+    INSERT INTO equipped (character_id, item_id)
+    SELECT character_id, item_id
+    FROM inventory
+    WHERE inventory_id = inventory_item_id;
+    
+    DELETE FROM inventory
+    WHERE inventory_id = inventory_item_id;
+END;
+//
+
+CREATE PROCEDURE unequip(IN equipped_item_id INT)
+BEGIN
+    DECLARE inventory_item_id INT;
+    
+    SELECT item_id INTO inventory_item_id
+    FROM equipped
+    WHERE equipped_id = equipped_item_id;
+    
+    INSERT INTO inventory (character_id, item_id)
+    SELECT character_id, item_id
+    FROM equipped
+    WHERE equipped_id = equipped_item_id;
+    
+    DELETE FROM equipped
+    WHERE equipped_id = equipped_item_id;
+END;
+//
+
+CREATE PROCEDURE set_winners(IN team_id INT)
```

Comment:
This PROCEDURE can't be verified due to another bug (see automated checks feedback). The intent of this PROCEDURE is to use a CURSOR to exercise those skills (that is part of the reason we never went over combining INSERT and SELECT up until this point). The provided solution is valid for a data set that is this small, but it doesn't scale to larger tables.

Reply:
The intent of this PROCEDURE is to use a CURSOR to exercise those skills (that is part of the reason we never went over combining INSERT and SELECT up until this point). The provided solution is valid for a data set that is this small, but it doesn't scale to larger tables.